## Data Loading and Cleaning:
-------

The Data Loading and Cleaning process outlines how the data was prepared for analysis. This includes, but is not limited to: Identifying and removing/imputing null values, identifying/removing duplicate data, Cleaning text data ect. 

##### Data Cleaning and Loading overview:
- Concatenating individual CSV files into a single data data frame.
- Exploring the data 
- Cleaning the data

##### Imports:
-------

In [1]:
import pandas as pd # Pandas imported and given the alias 'pd'
import numpy as np # numpy imported and given the alias 'np'. 
import matplotlib.pyplot as plt # matplotlib imported and given the alias 'plt'. Used for plotting graphs
import seaborn as sns 
import matplotlib as mlp

##### loading Datasets: 


In [2]:
# Reading the csv files
gbd = pd.read_csv('../data/youtube-dataset/GBvideos.csv')
frd = pd.read_csv('../data/youtube-dataset/FRvideos.csv')
usd = pd.read_csv('../data/youtube-dataset/USvideos.csv') 
cad = pd.read_csv('../data/youtube-dataset/CAvideos.csv') 
ded = pd.read_csv('../data/youtube-dataset/DEvideos.csv') 


KeyboardInterrupt: 

A total of 5 csv files will be used for this project. Each csv file presents YouTube statistics from a different country. The csv files will be merged later on to create a single data frame. Merging the csv files will make analysing the data easier during the Process of EDA. It is important to not that creating a data frame with large dimensions could make it difficult to fit a model on the data by increasing compuation times. However, working with a sample of this dataset could be utilized to mitigate this risk. 

Choosing to use global data as opposed to data from a singular country is useful for understanding how location/the video's origin can affect the amount of views a video recieves. This could be assessed futher by looking at country specific entries within the data eg: languages.

##### Checking individual entries:



In [ ]:
gbd.head() #Great Britain data set

video_id trending_date  \
0  Jw1Y-zhQURU      17.14.11   
1  3s1rvMFUweQ      17.14.11   
2  n1WpP7iowLc      17.14.11   
3  PUTEiSjKwJU      17.14.11   
4  rHwDegptbI4      17.14.11   

                                               title  \
0      John Lewis Christmas Ad 2017 - #MozTheMonster   
1          Taylor Swift: …Ready for It? (Live) - SNL   
2         Eminem - Walk On Water (Audio) ft. Beyoncé   
3  Goals from Salford City vs Class of 92 and Fri...   
4  Dashcam captures truck's near miss with child ...   

                channel_title  category_id              publish_time  \
0                  John Lewis           26  2017-11-10T07:38:29.000Z   
1         Saturday Night Live           24  2017-11-12T06:24:44.000Z   
2                  EminemVEVO           10  2017-11-10T17:00:03.000Z   
3  Salford City Football Club           17  2017-11-13T02:30:38.000Z   
4            Cute Girl Videos           25  2017-11-13T01:45:13.000Z   

                                                tags     views   likes  \
0  christmas|"john lewis christmas"|"john lewis"|...   7224515   55681   
1  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...   1053632   25561   
2  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787420   
3  Salford City FC|"Salford City"|"Salford"|"Clas...     27833     193   
4                                             [none]      9815      30   

   dislikes  comment_count                                  thumbnail_link  \
0     10247           9479  https://i.ytimg.com/vi/Jw1Y-zhQURU/default.jpg   
1      2294           2757  https://i.ytimg.com/vi/3s1rvMFUweQ/default.jpg   
2     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
3        12             37  https://i.ytimg.com/vi/PUTEiSjKwJU/default.jpg   
4         2             30  https://i.ytimg.com/vi/rHwDegptbI4/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  Click here to continue the story and make your...  
1  Musical guest Taylor Swift performs …Ready for...  
2  Eminem's new track Walk on Water ft. Beyoncé i...  
3  Salford drew 4-4 against the Class of 92 and F...  
4  Dashcam captures truck's near miss with child ...

In [ ]:
frd.head() #France data set 

video_id trending_date  \
0  Ro6eob0LrCY      17.14.11   
1  Yo84eqYwP98      17.14.11   
2  ceqntSXE-10      17.14.11   
3  WuTFI5qftCE      17.14.11   
4  ee6OFs8TdEg      17.14.11   

                                               title        channel_title  \
0           Malika LePen : Femme de Gauche - Trailer  Le Raptor Dissident   
1  LA PIRE PARTIE ft Le Rire Jaune, Pierre Croce,...              Le Labo   
2  DESSINS ANIMÉS FRANÇAIS VS RUSSES 2 - Daniil...      Daniil le Russe   
3                    PAPY GRENIER - METAL GEAR SOLID    Joueur Du Grenier   
4  QUI SAUTERA LE PLUS HAUT ? (VÉLO SKATE ROLLER ...    Aurelien Fontenoy   

   category_id              publish_time  \
0           24  2017-11-13T17:32:55.000Z   
1           24  2017-11-12T15:00:02.000Z   
2           23  2017-11-13T17:00:38.000Z   
3           20  2017-11-12T17:00:02.000Z   
4           17  2017-11-13T16:30:03.000Z   

                                                tags   views  likes  dislikes  \
0  Raptor"|"Dissident"|"Expliquez"|"moi"|"cette"|...  212702  29282      1108   
1                                             [none]  432721  14053       576   
2      cartoon"|"pokémon"|"école"|"ours"|"мультфильм  482153  76203       477   
3  Papy grenier"|"Metal Gear Solid"|"PS1"|"Tirage...  925222  85016       550   
4  vélo"|"vtt"|"bmx"|"freestyle"|"bike"|"mtb"|"di...  141695   8091        72   

   comment_count                                  thumbnail_link  \
0           3817  https://i.ytimg.com/vi/Ro6eob0LrCY/default.jpg   
1           1161  https://i.ytimg.com/vi/Yo84eqYwP98/default.jpg   
2           9580  https://i.ytimg.com/vi/ceqntSXE-10/default.jpg   
3           4303  https://i.ytimg.com/vi/WuTFI5qftCE/default.jpg   
4            481  https://i.ytimg.com/vi/ee6OFs8TdEg/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  Dimanche.\n18h30.\nSoyez présents pour la vidé...  
1  Le jeu de société: https://goo.gl/hhG1Ta\n\nGa...  
2  Une nouvelle dose de dessins animés français e...  
3  Nouvel ,épisode de Papy Grenier ! Ce mois-ci o...  
4  Sauts à plus de 4 mètres de haut dans un tramp...

In [ ]:
usd.head() #USA data set

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   

   category_id              publish_time  \
0           22  2017-11-13T17:13:01.000Z   
1           24  2017-11-13T07:30:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-13T11:00:04.000Z   
4           24  2017-11-12T18:01:41.000Z   

                                                tags    views   likes  \
0                                    SHANtell martin   748374   57527   
1  last week tonight trump presidency|"last week ...  2418783   97185   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434  146033   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168   10172   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731  132235   

   dislikes  comment_count                                  thumbnail_link  \
0      2966          15954  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1      6146          12703  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3       666           2146  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  Today we find out if Link is a Nickelback amat...  
4  I know it's been a while since we did this sho...

In [ ]:
cad.head() #Canada data set

video_id trending_date  \
0  n1WpP7iowLc      17.14.11   
1  0dBIkQ4Mz1M      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  d380meD0W0M      17.14.11   
4  2Vv-BfVoq4g      17.14.11   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   
3                           I Dare You: GOING BALD!?      nigahiga   
4        Ed Sheeran - Perfect (Official Music Video)    Ed Sheeran   

   category_id              publish_time  \
0           10  2017-11-10T17:00:03.000Z   
1           23  2017-11-13T17:00:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-12T18:01:41.000Z   
4           10  2017-11-09T11:04:14.000Z   

                                                tags     views    likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434   146035   
3  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095828   132239   
4  edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...  33523622  1634130   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
4     21082          85067  https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  I know it's been a while since we did this sho...  
4  🎧: https://ad.gt/yt-perfect\n💰: https://atlant...

## Merging Data-Sets:
------


##### Concatenating individual datasets:

In [ ]:
df = pd.concat([usd,frd,gbd,cad,ded]) # concatenating the different dataframes into one dataframe
print(df.shape)

(202310, 16)


##### Resetting the dataframe index

In [ ]:
df.reset_index(inplace=True,drop=True)
print(df.shape)
df.tail(1)

(202310, 16)


video_id trending_date  \
202309  go-F6xvezAM      18.14.06   

                                                    title       channel_title  \
202309  Гироскутер - Азбука Уральских Пельменей Б - Ур...  Уральские Пельмени   

        category_id              publish_time  \
202309           23  2018-06-13T15:02:15.000Z   

                                                     tags   views  likes  \
202309  Гироскутер|"уральские пельмени гироскутер"|"мя...  316328  11394   

        dislikes  comment_count  \
202309       352            550   

                                        thumbnail_link  comments_disabled  \
202309  https://i.ytimg.com/vi/go-F6xvezAM/default.jpg              False   

        ratings_disabled  video_error_or_removed  \
202309             False                   False   

                                              description  
202309  Популярный номер из нового шоу Азбука Уральски...

The data frame index was reset after merging to ensure that the indexing is consistent across the dataframe. Failure to reset the column index could result in overlapping indexes which could interfer with analysis on the data.

The data now contains: (202310 rows and 16 columns)

##### Checking the concatenated data:

In [ ]:
df.head() #Displays the first five rows of data (Checking for data format consistency ect.)

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   

   category_id              publish_time  \
0           22  2017-11-13T17:13:01.000Z   
1           24  2017-11-13T07:30:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-13T11:00:04.000Z   
4           24  2017-11-12T18:01:41.000Z   

                                                tags    views   likes  \
0                                    SHANtell martin   748374   57527   
1  last week tonight trump presidency|"last week ...  2418783   97185   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434  146033   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168   10172   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731  132235   

   dislikes  comment_count                                  thumbnail_link  \
0      2966          15954  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1      6146          12703  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3       666           2146  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  Today we find out if Link is a Nickelback amat...  
4  I know it's been a while since we did this sho...

In [ ]:
df.tail() #Displays the first five rows of data (Checking for data format consistency ect.)

video_id trending_date  \
202305  fn5WNxy-Wcw      18.14.06   
202306  zAFv43lxqHE      18.14.06   
202307  zSXG5I6Y2fA      18.14.06   
202308  5d115sePmaU      18.14.06   
202309  go-F6xvezAM      18.14.06   

                                                    title  \
202305  KINGDOM HEARTS III – E3 2018 Pirates of the Ca...   
202306                                     YMS: The Visit   
202307  Ungut umgeschult – Grünwald als Ersthelfer am ...   
202308  Assassin's Creed Odyssey: E3 2018 Welt-Enthüll...   
202309  Гироскутер - Азбука Уральских Пельменей Б - Ур...   

                  channel_title  category_id              publish_time  \
202305           Kingdom Hearts           20  2018-06-12T01:54:02.000Z   
202306     YourMovieSucksDOTorg           24  2018-06-13T21:58:43.000Z   
202307  Grünwald Freitagscomedy           24  2018-06-12T10:01:28.000Z   
202308      Assassin's Creed DE           20  2018-06-11T21:16:55.000Z   
202309       Уральские Пельмени           23  2018-06-13T15:02:15.000Z   

                                                     tags    views  likes  \
202305  Kingdom Hearts|"KH3"|"Kingdom Hearts 3"|"Pirat...  1394530  46778   
202306                                             [none]   139733  11155   
202307  Günter Grünwald|"Grünwald Freitagscomedy"|"Gün...    26054    364   
202308  Assassin's Creed|"Assassins Creed"|"Assassin's...  1139198  14900   
202309  Гироскутер|"уральские пельмени гироскутер"|"мя...   316328  11394   

        dislikes  comment_count  \
202305       501           9878   
202306       119           1968   
202307        11              8   
202308      1421           1587   
202309       352            550   

                                        thumbnail_link  comments_disabled  \
202305  https://i.ytimg.com/vi/fn5WNxy-Wcw/default.jpg              False   
202306  https://i.ytimg.com/vi/zAFv43lxqHE/default.jpg              False   
202307  https://i.ytimg.com/vi/zSXG5I6Y2fA/default.jpg              False   
202308  https://i.ytimg.com/vi/5d115sePmaU/default.jpg              False   
202309  https://i.ytimg.com/vi/go-F6xvezAM/default.jpg              False   

        ratings_disabled  video_error_or_removed  \
202305             False                   False   
202306             False                   False   
202307             False                   False   
202308             False                   False   
202309             False                   False   

                                              description  
202305  Find out more about Kingdom Hearts 3: https://...  
202306  Patreon: http://www.patreon.com/YMSTwitch: htt...  
202307  Günter versucht sich als Ersthelfer bei einem ...  
202308  Vom verstoßenen Söldner zum legendären Helden,...  
202309  Популярный номер из нового шоу Азбука Уральски...

Df.head and Df.tail have been used to show that the data entries are consistent throughout the dataframe. Each column contains the  appropriate data entries. 

Initial Observations:
- trending_date: Date format will need to be converted to date/time so it can be used for modelling.
- All date columns will be turned to numerical values and broken down into: month, day, year for modelling
- We could choose to keep non english text as column entries because they can reveal country specific trends about the global appeal of each country (Views)
- We Will need to Vectorize text data eg: tags and title so that they can be used for modelling
- publish_time and trending_date contain different date formats - potentially drop one column - multicolinearity
- Category id is numerical and provides little understandable information about the specific category - their corresponding values could be extracted through the data set's JSON Files to create a new names column.
- The distribution of numerical columns will be plotted to explore relationships: views, likes, dislikes,comment_count.
- Each video has a unique video id.


## Exploring the data:
----------

In [ ]:
df.shape # Provides the shape of the dataframe (Rows and Columns)

(202310, 16)

In [ ]:
df.columns # Provides all of the column names of the Data Frame.

Index(['video_id', 'trending_date', 'title', 'channel_title', 'category_id',
       'publish_time', 'tags', 'views', 'likes', 'dislikes', 'comment_count',
       'thumbnail_link', 'comments_disabled', 'ratings_disabled',
       'video_error_or_removed', 'description'],
      dtype='object')

In [ ]:
df.dtypes # This shows what datatype has been recorded for each individual columns.

video_id                  object
trending_date             object
title                     object
channel_title             object
category_id                int64
publish_time              object
tags                      object
views                      int64
likes                      int64
dislikes                   int64
comment_count              int64
thumbnail_link            object
comments_disabled           bool
ratings_disabled            bool
video_error_or_removed      bool
description               object
dtype: object

Most of the listed data types are correct. However, the 'publish_time' and 'trending_date' have been incorrectly listed as objects (They will need to be converted later on)

##### Number of unique values in id columns:

In [ ]:
df["category_id"].nunique()

18

There are a total of 18 different categories within the data. However, df.head has revealed that they are numerical, so category names will need to be obtained to gain more understanding of the data. 

In [ ]:
df["video_id"].nunique()

79408

There are a total of 202310 rows within the dataframe. However, nunique() has revealed that there are only 79408 unique video ids. This is important because it reveals that some videos repeat across the data from different countries.

In [ ]:
df.info() # Shows information about the data frame which, providing a summary of the data.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202310 entries, 0 to 202309
Data columns (total 16 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   video_id                202310 non-null  object
 1   trending_date           202310 non-null  object
 2   title                   202310 non-null  object
 3   channel_title           202310 non-null  object
 4   category_id             202310 non-null  int64 
 5   publish_time            202310 non-null  object
 6   tags                    202310 non-null  object
 7   views                   202310 non-null  int64 
 8   likes                   202310 non-null  int64 
 9   dislikes                202310 non-null  int64 
 10  comment_count           202310 non-null  int64 
 11  thumbnail_link          202310 non-null  object
 12  comments_disabled       202310 non-null  bool  
 13  ratings_disabled        202310 non-null  bool  
 14  video_error_or_removed  202310 non-n

In [ ]:
df.describe() # A statistical summary of the data.( mean, standard deviation, count ect.) 

category_id         views         likes      dislikes  comment_count
count  202310.000000  2.023100e+05  2.023100e+05  2.023100e+05   2.023100e+05
mean       19.712412  2.053181e+06  5.682284e+04  3.067639e+03   6.177626e+03
std         7.359156  9.412473e+06  2.078183e+05  2.859958e+04   3.145332e+04
min         1.000000  2.230000e+02  0.000000e+00  0.000000e+00   0.000000e+00
25%        17.000000  7.517425e+04  1.447000e+03  6.700000e+01   2.090000e+02
50%        23.000000  3.091290e+05  7.603000e+03  2.900000e+02   9.240000e+02
75%        24.000000  1.103690e+06  3.224475e+04  1.152000e+03   3.520000e+03
max        44.000000  4.245389e+08  5.613827e+06  1.944971e+06   1.626501e+06

In [ ]:
df.describe().T # A statistical summary of the data - columns repositioned.

count          mean           std    min       25%  \
category_id    202310.0  1.971241e+01  7.359156e+00    1.0     17.00   
views          202310.0  2.053181e+06  9.412473e+06  223.0  75174.25   
likes          202310.0  5.682284e+04  2.078183e+05    0.0   1447.00   
dislikes       202310.0  3.067639e+03  2.859958e+04    0.0     67.00   
comment_count  202310.0  6.177626e+03  3.145332e+04    0.0    209.00   

                    50%         75%          max  
category_id        23.0       24.00         44.0  
views          309129.0  1103690.50  424538912.0  
likes            7603.0    32244.75    5613827.0  
dislikes          290.0     1152.00    1944971.0  
comment_count     924.0     3520.00    1626501.0

In [ ]:
cat_cols = df.select_dtypes(include=['object']).columns.to_list() # selects all the category columns
num_cols = df.select_dtypes(include=['float64', 'bool']).columns.to_list() # selects all the numeric columns

In [ ]:
print(cat_cols)
print(num_cols)

['video_id', 'trending_date', 'title', 'channel_title', 'publish_time', 'tags', 'thumbnail_link', 'description']
['comments_disabled', 'ratings_disabled', 'video_error_or_removed']


##### Identifying numerical and categorical columns:

In [ ]:
df.select_dtypes('object').columns # Provides a list of non- numeric columns in the data - listed as object.

Index(['video_id', 'trending_date', 'title', 'channel_title', 'publish_time',
       'tags', 'thumbnail_link', 'description'],
      dtype='object')

In [ ]:
df.select_dtypes('int64','bool').columns # Provides a list of numeric columns in the data - listed as bool and int64.


Index(['category_id', 'views', 'likes', 'dislikes', 'comment_count'], dtype='object')


## Data Cleaning:
---------
- Cleaning the data to handle missing values, and inconsistencies. 
-  Data cleaning ensures the data is ready for analysis.

##### Duplicate data:

In [ ]:
df.duplicated().sum() # This provides the sum of all duplicated columns in the Data Frame 

18083

Df.duplicated has revealed that there are 18083 duplicated rows across the data set. This is a large number of duplicated values. Considering the nature of this dataset 'Trending YouTube videos', it is possible that the data contains video entries that have trended across multiple countries. The decision to remove this data could limit the how well our model will be able to predict which features are the most statistically significant to the number of views.

In [ ]:
df[df.duplicated()] # to find all the duplicates data in the Data Frame.

video_id trending_date  \
34899   QBL8IRJ5yHU      18.15.05   
34900   t4pRQ0jn23Q      18.15.05   
34901   j4KvrAUjn6c      18.15.05   
34902   MAjY8mCTXWk      18.15.05   
34903   xhs8tf1v__w      18.15.05   
...             ...           ...   
202300  DGzy8FE1Rhk      18.14.06   
202301  HyqTJpG_JbE      18.14.06   
202303  C9DCE3k2lWA      18.14.06   
202304  KYke3FFiyk4      18.14.06   
202306  zAFv43lxqHE      18.14.06   

                                                    title  \
34899   Why I'm So Scared (being myself and crying too...   
34900   YoungBoy Never Broke Again Goes Sneaker Shoppi...   
34901          WE MADE OUR MOM CRY...HER DREAM CAME TRUE!   
34902   周杰倫 Jay Chou【不愛我就拉倒 If You Don't Love Me, It's...   
34903   Terry Crews Answers the Web's Most Searched Qu...   
...                                                   ...   
202300                             Shawn Mendes - Nervous   
202301  هذه هي الدول التي لم تصوت على المغرب لإستضافة ...   
202303  الرئيس أحمد أحمد يصرح بأنه تلقى تهديدات ويتحسر...   
202304  Crime Patrol Dial 100 - Ep 796 - Full Episode ...   
202306                                     YMS: The Visit   

               channel_title  category_id              publish_time  \
34899          grav3yardgirl           26  2018-05-14T19:00:01.000Z   
34900                Complex           24  2018-05-14T14:00:03.000Z   
34901       Lucas and Marcus           24  2018-05-13T18:03:56.000Z   
34902        杰威爾音樂 JVR Music           10  2018-05-14T15:59:47.000Z   
34903                  WIRED           24  2018-05-14T16:00:29.000Z   
...                      ...          ...                       ...   
202300       ShawnMendesVEVO           10  2018-06-11T16:00:03.000Z   
202301      DailyProFootball           17  2018-06-13T12:05:22.000Z   
202303             Shahid TV           17  2018-06-13T19:31:05.000Z   
202304             SET India           24  2018-06-13T13:54:47.000Z   
202306  YourMovieSucksDOTorg           24  2018-06-13T21:58:43.000Z   

                                                     tags    views   likes  \
34899   beauty|"how to"|"makeup"|"howto"|"style"|"fash...  1469627  188652   
34900   sneakerhead|"complex"|"complex originals"|"sne...  1199587   49709   
34901   Lucas and Marcus|"Marcus and Lucas"|"Dobre"|"D...  3906727   77378   
34902   周杰倫|"Jay"|"Chou"|"周董"|"周杰伦"|"周傑倫"|"杰威尔"|"周周"|"...   916128   40485   
34903   autocomplete|"deadpool 2"|"google autocomplete...   343967   16988   
...                                                   ...      ...     ...   
202300  Shawn|"Mendes"|"Nervous"|"Island"|"Records"|"Pop"  4986664  518240   
202301  المغرب 2026|"morocco 2026"|"كأس العالم 2026"|"...   184156     504   
202303  الرئيس|"أحمد"|"يصرح"|"بأنه"|"تلقى"|"تهديدات"|"...   184242    1440   
202304  true events|"sony entertainment channel"|"cons...    54395     285   
202306                                             [none]   139733   11155   

        dislikes  comment_count  \
34899       3124          33032   
34900       2380           7261   
34901      12160          15874   
34902       1042           4746   
34903        132           1308   
...          ...            ...   
202300      5215          34466   
202301       217           1368   
202303        97            470   
202304        61             37   
202306       119           1968   

                                        thumbnail_link  comments_disabled  \
34899   https://i.ytimg.com/vi/QBL8IRJ5yHU/default.jpg              False   
34900   https://i.ytimg.com/vi/t4pRQ0jn23Q/default.jpg              False   
34901   https://i.ytimg.com/vi/j4KvrAUjn6c/default.jpg              False   
34902   https://i.ytimg.com/vi/MAjY8mCTXWk/default.jpg              False   
34903   https://i.ytimg.com/vi/xhs8tf1v__w/default.jpg              False   
...                                                ...                ...   
202300  https://i.ytimg.com/vi/DGzy8FE1Rhk/default.jpg              False   
202301

In [ ]:
#df.T.duplicated() # Checks to see if any of the columns are duplicated

There are no duplicated columns in the dataframe.

##### Null values:

In [ ]:
df.isna().sum() # checks for null values in all of the columns.

video_id                     0
trending_date                0
title                        0
channel_title                0
category_id                  0
publish_time                 0
tags                         0
views                        0
likes                        0
dislikes                     0
comment_count                0
thumbnail_link               0
comments_disabled            0
ratings_disabled             0
video_error_or_removed       0
description               6942
dtype: int64

In [ ]:
df.isna().mean()*100 # Calculates the mean percentage of null values for each column.

video_id                  0.000000
trending_date             0.000000
title                     0.000000
channel_title             0.000000
category_id               0.000000
publish_time              0.000000
tags                      0.000000
views                     0.000000
likes                     0.000000
dislikes                  0.000000
comment_count             0.000000
thumbnail_link            0.000000
comments_disabled         0.000000
ratings_disabled          0.000000
video_error_or_removed    0.000000
description               3.431368
dtype: float64

In [ ]:
df.isna().sum().sum() # this collects the total sum of null values across the Data Frame. 

6942

There is a total of 6942 null values within the description column. All other columns contain no null values. Majority of the data within this dataset contains unique data entries, so imputing these values from other columns within the dataset would not be worthwhile because it would take too long. These values would most likely need to be webscaped. However, I do not think that webscraping is needed becuase the data relating to this column could be readily available within the remaining text columns eg: tags. For now, the null values will be dropped.

In [ ]:
df_dropnul = df.copy() # new data frame to make changes on

In [ ]:
df_dropnul.dropna(inplace=True) # dropping null values

In [ ]:
df_dropnul[df_dropna['tags'] == '[none]'] # Missing values within the tags

video_id trending_date  \
97      xfmipNU4Odc      17.14.11   
133     X7flefV8tec      17.14.11   
136     5x1FAiIq_pQ      17.14.11   
178     JuP1Z8xpRb8      17.14.11   
180     1640fZpYBSY      17.14.11   
...             ...           ...   
202199  pGxPSPqwdiw      18.14.06   
202227  KAyj5Xm1C64      18.14.06   
202247  VdOGUFr3glA      18.14.06   
202296  LR308Yr8tsg      18.14.06   
202306  zAFv43lxqHE      18.14.06   

                                                    title  \
97      Edna's registered owner thought she was dead f...   
133     President Bill Clinton On Dictators, Democracy...   
136                      Alicia Keys - When You Were Gone   
178                         Brian Justin Crum - Wild Side   
180               I love the Price is Right! Wooo! -Kevin   
...                                                   ...   
202199      Një shtepi u inagurua ne qytetin e Vushtrrise   
202227  [ENG SUB] BTS PROM PARTY 2018 Intro + 2nd Gran...   
202247  لحظة تتويج ملف التلاتي بكأس العالم 2026 -خسارة...   
202296               LIVE NOW ! - 68th FIFA Congress 2018   
202306                                     YMS: The Visit   

                                  channel_title  category_id  \
97      Hope For Paws - Official Rescue Channel           15   
133                                   Team Coco           24   
136                                 Alicia Keys           10   
178                             BrianJustinCrum           24   
180                                   Anaki Abo           22   
...                                         ...          ...   
202199                      Jetimat e Ballkanit           22   
202227                             DaisyxBTS 07           24   
202247                   Sahifa-Tv قناة الصحيفة           24   
202296                                   FIFATV           17   
202306                     YourMovieSucksDOTorg           24   

                    publish_time    tags   views  likes  dislikes  \
97      2017-11-10T18:02:04.000Z  [none]  284666  16396        81   
133     2017-11-09T02:37:49.000Z  [none]  366180   4364      4448   
136     2017-11-09T15:49:21.000Z  [none]   95944   1354       181   
178     2017-11-09T17:01:01.000Z  [none]   27010   1666        36   
180     2017-11-07T18:54:39.000Z  [none]  358597   1211        72   
...                          ...     ...     ...    ...       ...   
202199  2018-06-13T18:01:37.000Z  [none]   44577    658        16   
202227  2018-06-13T12:51:23.000Z  [none]  449418  24806        93   
202247  2018-06-13T11:00:10.000Z  [none]  429897   1915      1021   
202296  2018-06-13T12:24:25.000Z  [none]  312470   2752       412   
202306  2018-06-13T21:58:43.000Z  [none]  139733  11155       119   

        comment_count                                  thumbnail_link  \
97                949  https://i.ytimg.com/vi/xfmipNU4Odc/default.jpg   
133              1997  https://i.ytimg.com/vi/X7flefV8tec/default.jpg   
136               117  https://i.ytimg.com/vi/5x1FAiIq_pQ/default.jpg   
178               150  https://i.ytimg.com/vi/JuP1Z8xpRb8/default.jpg   
180               593  https://i.ytimg.com/vi/1640fZpYBSY/default.jpg   
...               ...                                             ...   
202199             63  https://i.ytimg.com/vi/pGxPSPqwdiw/default.jpg   
202227            974  https://i.ytimg.com/vi/KAyj5Xm1C64/default.jpg   
202247           1635  https://i.ytimg.com/vi/VdOGUFr3glA/default.jpg   
202296            133  https://i.ytimg.com/vi/LR308Yr8tsg/default.jpg   
202306           1968  https://i.ytimg.com/vi/zAFv43lxqHE/default.jpg   

        comments_disabled  ratings_disabled  video_error_or_removed  \
97                  False             False                   False   
133                 False             False                   False   
136                 False             False                   False   
178                 False             False                   False   
180

Missing values have also been detected within the tags column. However, they will be left alone because they have already been listed as 'none'. It is unlikely that leaving these values will have a negative impact on predicting 'views' because some of this information could also be found within the description and title columns.

##### Closer look at the text data: Description

In [ ]:
df_dropnul['description'].head(10) #first 10 rows of data from the desciption column.

0    SHANTELL'S CHANNEL - https://www.youtube.com/s...
1    One year after the presidential election, John...
2    WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...
3    Today we find out if Link is a Nickelback amat...
4    I know it's been a while since we did this sho...
5    Using the iPhone for the past two weeks -- her...
6    Embattled Alabama Senate candidate Roy Moore (...
7    Ice Cream Pint Combination Lock - http://amzn....
8    Inspired by the imagination of P.T. Barnum, Th...
9    For now, at least, we have better things to wo...
Name: description, dtype: object

The description can be considered an important column because it contains the most text data. From looking at this column it is clear that a large amount of unuseful data eg: special characters that will need to be removed. Although cleaning this data is an important step, most of the the unuseful data will be filtered out by the CountVectorizer when the data is vectorized.


## Cleaning Text Data 
-------

##### Text formatting to lowercase:
Converting description, tags and title to lower case

In [ ]:
df_transform = df_dropnul.copy()

In [ ]:
# converting all text to lower case, replacing the original column values.

df_transform['tags_lower'] = df_transform['tags'].apply(lambda x: x.lower()) #convert tags to lower case
df_transform['title_lower'] = df_transform['title'].apply(lambda x: x.lower()) # converts title to lowercase 
df_transform['description_lower'] = df_transform['description'].apply(lambda x: str(x).lower()) # converts description to lowercase 



In [ ]:
print(df_transform['tags_lower']) #confirming the column has been converted to lowercase.

0                                           shantell martin
1         last week tonight trump presidency|"last week ...
2         racist superman|"rudy"|"mancuso"|"king"|"bach"...
3         rhett and link|"gmm"|"good mythical morning"|"...
4         ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...
                                ...                        
202305    kingdom hearts|"kh3"|"kingdom hearts 3"|"pirat...
202306                                               [none]
202307    günter grünwald|"grünwald freitagscomedy"|"gün...
202308    assassin's creed|"assassins creed"|"assassin's...
202309    гироскутер|"уральские пельмени гироскутер"|"мя...
Name: tags_lower, Length: 195368, dtype: object


In [ ]:
print(df_transform['title_lower']) #confirming the column has been converted to lowercase.

0                        we want to talk about our marriage
1         the trump presidency: last week tonight with j...
2         racist superman | rudy mancuso, king bach & le...
3                          nickelback lyrics: real or fake?
4                                  i dare you: going bald!?
                                ...                        
202305    kingdom hearts iii – e3 2018 pirates of the ca...
202306                                       yms: the visit
202307    ungut umgeschult – grünwald als ersthelfer am ...
202308    assassin's creed odyssey: e3 2018 welt-enthüll...
202309    гироскутер - азбука уральских пельменей б - ур...
Name: title_lower, Length: 195368, dtype: object


In [ ]:
print(df_transform['description_lower']) #confirming the column has been converted to lowercase.

0         shantell's channel - https://www.youtube.com/s...
1         one year after the presidential election, john...
2         watch my previous video ▶ \n\nsubscribe ► http...
3         today we find out if link is a nickelback amat...
4         i know it's been a while since we did this sho...
                                ...                        
202305    find out more about kingdom hearts 3: https://...
202306    patreon: http://www.patreon.com/ymstwitch: htt...
202307    günter versucht sich als ersthelfer bei einem ...
202308    vom verstoßenen söldner zum legendären helden,...
202309    популярный номер из нового шоу азбука уральски...
Name: description_lower, Length: 195368, dtype: object


## Text Cleaning:
------
Removing all numerical values and special characters from text columns: Title, Tags & Description.

In [ ]:
df_cleaningtext = df_transform.copy()

In [ ]:
# removing unicode character eg: hyperlinks, emojis punctuations
import re
pattern = r"(@\[A-Za-z0-9]+)|([^0-9A-Za-z \t])|(\w+:\/\/\S+)|^rt|http.+?"
replace = "" 

df_cleaningtext['tags_no_chars'] = df_cleaningtext['tags_lower'].apply(lambda x: re.sub(pattern, replace, x))
df_cleaningtext['description_no_chars'] = df_cleaningtext['description_lower'].apply(lambda x: re.sub(pattern, replace, x))
df_cleaningtext['title_no_chars'] = df_cleaningtext['title_lower'].apply(lambda x: re.sub(pattern, replace, x))


In [ ]:
# removing numbers dates from tags eg: '2002'
df_cleaningtext['tags_no_dates'] = df_cleaningtext['tags_no_chars'].apply(lambda x: re.sub(r'\b\d{4}\b', '', x)) #removing the date entries
df_cleaningtext['tags_no_numerics'] = df_cleaningtext['tags_no_dates'].apply(lambda x: re.sub(r'\d+', '', x)) #removing individual numbers

description = (filter(lambda x: not x.isdigit(), df['description']))
df_cleaningtext['tags_no_numerics']

0                                           shantell martin
1         last week tonight trump presidencylast week to...
2         racist supermanrudymancusokingbachracistsuperm...
3         rhett and linkgmmgood mythical morningrhett an...
4         ryanhigahigatvnigahigai dare youidyrhpcdaresno...
                                ...                        
202305    kingdom heartskhkingdom hearts pirates of the ...
202306                                                 none
202307    gnter grnwaldgrnwald freitagscomedygnter grnwa...
202308    assassins creedassassins creedassassins creed ...
202309                                                     
Name: tags_no_numerics, Length: 195368, dtype: object

In [ ]:
#to remove single numerical values in the tags column eg: '2'
df_cleaningtext['tags_clean'] = df_cleaningtext['tags_no_numerics'].str.replace(r'\d+', '')  #replaces numeric values with whitespace.

/var/folders/1_/mgb_spqs3b71d7zj3xh9r_6m0000gn/T/ipykernel_1185/1248817232.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  df_cleaningtext['tags_clean'] = df_cleaningtext['tags_no_numerics'].str.replace(r'\d+', '')  #replaces numeric values with whitespace.


In [ ]:
df_cleaningtext['tags_clean'] #checking  the data entries inside the tags column.

0                                           shantell martin
1         last week tonight trump presidencylast week to...
2         racist supermanrudymancusokingbachracistsuperm...
3         rhett and linkgmmgood mythical morningrhett an...
4         ryanhigahigatvnigahigai dare youidyrhpcdaresno...
                                ...                        
202305    kingdom heartskhkingdom hearts pirates of the ...
202306                                                 none
202307    gnter grnwaldgrnwald freitagscomedygnter grnwa...
202308    assassins creedassassins creedassassins creed ...
202309                                                     
Name: tags_clean, Length: 195368, dtype: object

- Percentage of null values is small across the dataframe so it is not significant. 
- Although the contents of this column could be found in other column, it may be useful with understanding the category_id.
- The null values in these columns will be imputed through webscraping descriptions online. 

## Numerical Columns:
----

In [ ]:
numerical_columns = df_cleaningtext.select_dtypes(include=['float64', 'int64'])
numerical_columns

category_id    views   likes  dislikes  comment_count
0                22   748374   57527      2966          15954
1                24  2418783   97185      6146          12703
2                23  3191434  146033      5339           8181
3                24   343168   10172       666           2146
4                24  2095731  132235      1989          17518
...             ...      ...     ...       ...            ...
202305           20  1394530   46778       501           9878
202306           24   139733   11155       119           1968
202307           24    26054     364        11              8
202308           20  1139198   14900      1421           1587
202309           23   316328   11394       352            550

[195368 rows x 5 columns]

In [ ]:
df2 = df_cleaningtext.copy() #column to make changes on - dropping the columns

In [ ]:
df2['comments_disabled'].value_counts()


False    191769
True       3599
Name: comments_disabled, dtype: int64

In [ ]:
df2['ratings_disabled'].value_counts()

False    193511
True       1857
Name: ratings_disabled, dtype: int64

In [ ]:
df2['video_error_or_removed'].value_counts()

False    195218
True        150
Name: video_error_or_removed, dtype: int64

The 'comments_disabled', 'ratings_disabled', 'video_error_or_removed' will be removed because they do not contain any usefull information to predict the number of 'views'. Furthermore, all the values in these binary columns are majority 'false' which means that these columns do not contain enough varied data for it to be meaninful/useful for the target.

In [ ]:
df2 = df.drop(columns=['comments_disabled', 'ratings_disabled', 'video_error_or_removed']) #dropping columns

In [ ]:
df2.head() # showing the first 5 columns to observe changes in the dataframe.

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   

   category_id              publish_time  \
0           22  2017-11-13T17:13:01.000Z   
1           24  2017-11-13T07:30:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-13T11:00:04.000Z   
4           24  2017-11-12T18:01:41.000Z   

                                                tags    views   likes  \
0                                    SHANtell martin   748374   57527   
1  last week tonight trump presidency|"last week ...  2418783   97185   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434  146033   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168   10172   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731  132235   

   dislikes  comment_count                                  thumbnail_link  \
0      2966          15954  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1      6146          12703  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3       666           2146  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  Today we find out if Link is a Nickelback amat...  
4  I know it's been a while since we did this sho...

##### Resetting Dataframe indexes:

In [ ]:
df2.reset_index(inplace=True,drop=True)
print(df2.shape)
df2.tail(1)

(202310, 13)


video_id trending_date  \
202309  go-F6xvezAM      18.14.06   

                                                    title       channel_title  \
202309  Гироскутер - Азбука Уральских Пельменей Б - Ур...  Уральские Пельмени   

        category_id              publish_time  \
202309           23  2018-06-13T15:02:15.000Z   

                                                     tags   views  likes  \
202309  Гироскутер|"уральские пельмени гироскутер"|"мя...  316328  11394   

        dislikes  comment_count  \
202309       352            550   

                                        thumbnail_link  \
202309  https://i.ytimg.com/vi/go-F6xvezAM/default.jpg   

                                              description  
202309  Популярный номер из нового шоу Азбука Уральски...

In [ ]:
set(df2['trending_date'].str[:3]) #resetting the date format

{'17.', '18.'}

##### Pickling the data frame:

In [ ]:
#pickling the dataframe

import joblib
joblib.dump(df2,'df_cleaning.pkl')

['df_cleaning.pkl']

## Summary:
